sql-plugin/src/main/scala/org/apache/spark/sql/rapids/predicates.scala

```diff
@@ -66,6 +67,125 @@ case class GpuAnd(left: Expression, right: Expression) extends CudfBinaryOperato
+    }
+    val gatherMap = withResource(prefixSumExclusive) { prefixSumExclusive =>
+      // for the entries in the gather map that do not represent valid
+      // values to be gathered, we change the value to -MAX_INT which
+      // will be treated as null values in the gather algorithm
+      withResource(Scalar.fromInt(Int.MinValue)) {
+        outOfBoundsFlag => predicate.ifElse(prefixSumExclusive, outOfBoundsFlag)
+      }
+    }
+    withResource(gatherMap) { _ =>
+      withResource(new Table(t.getBase)) { tbl =>
+        withResource(tbl.gather(gatherMap)) { gatherTbl =>
+          gatherTbl.getColumn(0).incRefCount()
+        }
+      }
+    }
+  }
+
+  protected def isAllFalse(col: GpuColumnVector): Boolean = {
+    assert(BooleanType == col.dataType())
+    if (col.getRowCount == col.numNulls()) {
+      // all nulls, and null values are false values here
+      return true
+    }
+    withResource(col.getBase.any()) { anyTrue =>
+      // null values are considered false values in this context
+      !anyTrue.getBoolean
+    }
+  }
+
+  /**
+   * When computing logical expressions on the CPU, the true and false
+   * expressions are evaluated lazily, meaning that the RHS expression
+   * of logical-AND is not evaluated when LHS is False. For logical-OR,
+   * RHS is not evaluated when LHS is True.
+   * This is important in the case where the expressions can have
+   * side-effects, such as throwing exceptions for invalid inputs.
+   *
+   * This method performs lazy evaluation on the GPU by first filtering
+   * the input batch a  where the LHS predicate is True.
+   * The RHS predicate is evaluated against these batches and then the
+   * results are combined back into a single batch using the gather
+   * algorithm.
+   */
+  private def columnarEvalWithSideEffects(batch: ColumnarBatch): Any = {
+    val leftExpr = left.asInstanceOf[GpuExpression]
+    val rightExpr = right.asInstanceOf[GpuExpression]
+    val colTypes = GpuColumnVector.extractTypes(batch)
+
+    withResource(GpuColumnVector.from(batch)) { tbl =>
+      withResource(GpuExpressionsUtils.columnarEvalToColumn(leftExpr, batch)) { lhsBool =>
+        if (isAllFalse(lhsBool)) {
+          withResource(GpuScalar.from(false, dataType)) { falseScalar =>
+            GpuColumnVector.from(falseScalar, lhsBool.getRowCount.toInt, BooleanType)
+          }
+        } else {
+          val rEval = withResource(filterBatch(tbl, lhsBool.getBase, colTypes)) { leftTrueBatch =>
+            rightExpr.columnarEval(leftTrueBatch)
+          }
+          val finalRet = withResourceIfAllowed(rEval) { _ =>
+            (lhsBool, rEval) match {
+              case (t: GpuColumnVector, f: GpuColumnVector) =>
+                withResource(gather(t.getBase, f)) { combinedVector =>
+                  t.getBase.and(combinedVector)
+                }
+              case (t: GpuColumnVector, f: GpuScalar) =>
+                t.getBase.and(f.getBase)         
+            }
+          }
```

Comment:
This can be simplied since `lhsBool` is of a fixed type, so need to match on that. We can also the [Pattern Matching Anonymous Function](https://www.scala-lang.org/files/archive/spec/2.13/08-pattern-matching.html#pattern-matching-anonymous-functions) syntax to make it a little more concise - this basically removes the need for the explicit `rEval match`.

```suggestion
          val finalRet = withResourceIfAllowed(rEval) {
            case r: GpuColumnVector =>
              withResource(gather(lhsBool.getBase, r)) { combinedVector =>
                lhsBool.getBase.and(combinedVector)
              }
            case r: GpuScalar =>
              lhsBool.getBase.and(r.getBase)
          }
```